### PR TITLE
(Merge-on-green) fix: add MOG tests, changed checking status of PR from merged to closed, and add logic to use native branch protection

### LINKED ISSUE #353: merge on green doesn't seem to handle closed PRs
There is a log of pull requests that were closed but not merged, see:

closed-not-merged

### PR DESCRIPTION
**Pull request by sofisl**
Fixes #353, #346

### CODE REVIEW

File: packages/merge-on-green/package.json [hunk 1]
```diff
@@ -27,7 +27,9 @@
   },
   "dependencies": {
     "@google-cloud/datastore": "^5.0.4",
-    "gcf-utils": "1.6.2"
+    "@types/sinon": "^7.5.2",
```
Comment (JustinBeckwith) [inline packages/merge-on-green/package.json]: Guessing this should be a dev-dependency (along with sinon)

File: packages/merge-on-green/src/merge-on-green.ts [hunk 2]
```diff
@@ -30,15 +30,25 @@ interface WatchPR {
   url: string;
 }
 
-handler.listPRs = async function listPRs(): Promise<WatchPR[]> {
+handler.getDatastore = async function getDatastore() {
   const query = datastore.createQuery(TABLE).order('created');
   const [prs] = await datastore.runQuery(query);
+  return [prs];
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-on-green.ts]: Do we want to return an array here?  May be nice to add an explicit return type so we know this was on purpose (if that's the case)

File: packages/merge-on-green/src/merge-on-green.ts [hunk 3]
```diff
@@ -30,15 +30,25 @@ interface WatchPR {
   url: string;
 }
 
-handler.listPRs = async function listPRs(): Promise<WatchPR[]> {
+handler.getDatastore = async function getDatastore() {
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-on-green.ts]: I'd love a jsdoc header that describes what this method does

File: packages/merge-on-green/src/merge-on-green.ts [hunk 4]
```diff
@@ -30,15 +30,25 @@ interface WatchPR {
...
   const query = datastore.createQuery(TABLE).order('created');
   const [prs] = await datastore.runQuery(query);
+  return [prs];
+};
+
+handler.listPRs = async function listPRs(): Promise<WatchPR[]> {
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-on-green.ts]: For each of these, would love a header with a 1-2 sentence description of what the function is used for.  Helpful for unsuspecting souls that wonder into the codebase later (me :))

File: packages/merge-on-green/src/merge-on-green.ts [hunk 5]
```diff
@@ -30,15 +30,25 @@ interface WatchPR {
...
+  const [prs] = await handler.getDatastore();
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-on-green.ts]: Yeah, if we're just destructuring the array here, no reason to return an array of results above.  We can remove the `[` and `]` here, and in the function that returns the results.

File: packages/merge-on-green/src/merge-on-green.ts [hunk 6]
```diff
@@ -30,15 +30,25 @@ interface WatchPR {
...
   const result: WatchPR[] = [];
   for (const pr of prs) {
     const created = new Date(pr.created).getTime();
     const now = new Date().getTime();
-    const url = pr[datastore.KEY].name;
     let state = 'continue';
+    let url;
+    if (pr[datastore.KEY] !== undefined) {
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-on-green.ts]: I think we can now say:

```js
const url = pr[datastore.KEY]?.name;
```

And it should have the same effect.

File: packages/merge-on-green/src/merge-on-green.ts (hunk 5 above)
Comment (sofisl) [reply packages/merge-on-green/src/merge-on-green.ts]: Thanks for taking a look, Justin! To be honest, I had some trouble with this function. We're technically returning [this](https://cloud.google.com/datastore/docs/reference/data/rpc/google.datastore.v1#google.datastore.v1.RunQueryResponse
) type, but it's actually just the QueryResultBatch part of that Response, since we're selecting an array. However, that data type is not exported, and TS won't let me convert it to my own type. I'm probably doing something wrong, I just can't quite figure out what it is! :)

File: packages/merge-on-green/src/merge-on-green.ts (hunk 2 above)
Comment (sofisl) [reply packages/merge-on-green/src/merge-on-green.ts]: See comment below...Would love your help on it if you have any ideas!

**Comment by sofisl** [comment]
@JustinBeckwith and @bcoe - I ended up adding the new branch logic we agreed on to this PR, since the point of this PR was to finalize adding tests, and the change in branch logic would have made me had to rewrite the tests completely anyways. Sorry for the long PR!!

File: packages/merge-on-green/src/merge-logic.ts [hunk 7]
```diff
@@ -141,166 +172,165 @@ mergeOnGreen.hasMOGLabel = async function hasMOGLabel(
...
+    ).data;
+    console.info(
+      `called getStatusi in ${Date.now() - start}ms ${owner}/${repo}, ${
+        data[0].context
+      }`
+    );
+    return data;
   } catch (err) {
-    console.error(err);
     return [];
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-logic.ts]: Same feedback as above, wondering if we should log the error

File: packages/merge-on-green/src/merge-logic.ts [hunk 8]
```diff
@@ -141,166 +172,165 @@ mergeOnGreen.hasMOGLabel = async function hasMOGLabel(
...
+  num: number
+): Promise<CheckStatus[]> {
+  const start = Date.now();
   try {
-    const configFile = (
-      await github.repos.getContents({
-        owner: 'googleapis',
-        repo: 'sloth',
-        path: 'repos.json',
+    const data = (
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-logic.ts]: You can shorten this to `const {data} = await github.repos.listStatusesForRef....`

File: packages/merge-on-green/src/merge-logic.ts [hunk 9]
```diff
@@ -141,166 +172,165 @@ mergeOnGreen.hasMOGLabel = async function hasMOGLabel(
...
+        per_page: 100,
+        page: num,
       })
-    ).data as { content?: string };
-    return JSON.parse(
-      Buffer.from(configFile.content as string, 'base64').toString('utf8')
-    ).repos;
+    ).data;
+    console.info(
+      `called getStatusi in ${Date.now() - start}ms ${owner}/${repo}, ${
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-logic.ts]: "getStatusi` typo :)

File: packages/merge-on-green/src/merge-logic.ts [hunk 10]
```diff
@@ -141,166 +172,165 @@ mergeOnGreen.hasMOGLabel = async function hasMOGLabel(
...
-      return checksByLanguage[language.language].requiredStatusChecks;
-    } else {
-      console.info(
-        `${owner}/${repo} does not have a corresponding language in sloth/repos.json`
-      );
-      return [];
+  headSha: string
+): Promise<CheckStatus[]> {
+  let results: CheckStatus[] = [];
+  for (let i = 0; i < 10; i++) {
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-logic.ts]: "10" is an oddly round number.  What's happening here?  Is that a concurrency limit or something?

File: packages/merge-on-green/src/merge-on-green.ts [hunk 11]
```diff
@@ -18,49 +18,90 @@ import { mergeOnGreen } from './merge-logic';
...
   number: number;
   repo: string;
   owner: string;
-  state: 'continue' | 'stop';
+  state: 'continue' | 'stop' | 'comment';
   url: string;
 }
 
-handler.listPRs = async function listPRs(): Promise<WatchPR[]> {
+// interface QueryResponse {
```
Comment (JustinBeckwith) [inline packages/merge-on-green/src/merge-on-green.ts]: Commented out interface - should remove

File: packages/merge-on-green/src/merge-on-green.ts [hunk 12]
```diff
@@ -18,49 +18,83 @@ import { mergeOnGreen } from './merge-logic';
...
     let state = 'continue';
+    const url = pr[datastore.KEY]?.name;
     //TODO: I'd prefer to not have a "list" method that has side effects - perhaps later refactor
     //this to do the list, then have an explicit loop over the returned WatchPR objects that removes the expired ones.
     if (now - created > MAX_TEST_TIME) {
       state = 'stop';
     }
+    if (
+      now - created < COMMENT_INTERVAL_HIGH &&
+      now - created > COMMENT_INTERVAL_LOW
```
Comment (bcoe) [inline packages/merge-on-green/src/merge-on-green.ts]: this addition is neat.

File: packages/merge-on-green/src/merge-logic.ts [hunk 13]
```diff
@@ -137,152 +149,115 @@ mergeOnGreen.hasMOGLabel = async function hasMOGLabel(
...
+        repo,
+        branch: 'master',
+      })
+    ).data.required_status_checks.contexts;
+    console.log(
+      `checking branch protection for ${owner}/${repo}: ${branchProtection}`
     );
+    return branchProtection;
+  } catch (err) {
+    console.log(`Error in getting branch protection: ${err}`);
```
Comment (bcoe) [inline packages/merge-on-green/src/merge-logic.ts]: we should be mindful about continuing to add context about what repo is having this error. Eventually we should pull this logic up a layer, so that any bot just gets it for free if they use a built in logger.
